### DISCUSSION TITLE
2020 November progress report

### DISCUSSION THREAD

**Discussion by FooBarWidget** [fullstaq-labs/fullstaq-ruby-server-edition#68] [Show and tell]
The focus in the past few months has been on Epic 4, whose goal is to improve contribution friendliness. As outlined in the [project vision](https://www.joyfulbikeshedding.com/blog/2020-05-15-why-fullstaq-ruby.html), we want Fullstaq Ruby to remain a sustainable and healthy project that doesn't rely on just a few key individuals.

In order to achieve that, we invest heavily into [release automation (a CI/CD system)](https://www.joyfulbikeshedding.com/blog/2020-05-05-fullstaq-ruby-epic-3.html) and organizational documentation (the [Development Handbook](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/blob/main/dev-handbook/README.md)). So that anybody can contribute to the project, with ease.

## Are We Contributor Friendly Yet?

It seems this investment is already paying off. Not long ago, @caius [contributed Ruby 2.7.2 support](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/pull/61). In his words:

> (PS: This project is awesome, and I love the included documentation. Really easy to get a sense of how it hangs together and works. Thank you very much for creating this project 👏.)

## Roadblocks

Of course, no journey is without roadblocks. Building release automation turns out to be harder than one might expect. The biggest issue we ran into recently, was the fact that we [ran into undocumented and unofficial Github Actions limits](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/issues/62).

Why did this happen? It's because of the sheer number of packages this project has to build. The number of packages grows exponentially with the number of distributions, Ruby versions and variants we support (see [Build workflow management](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/blob/main/dev-handbook/build-workflow-management.md)). We have to build so many packages, that Github Actions can't handle the sheer number of jobs and steps defined in the CI/CD system.

After talking to a Github Actions product owner, it was clear that this issue wouldn't be solved any time soon. After [evaluating our options](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/issues/62#issuecomment-704746817), it was decided that splitting the CI/CD system into multiple workflows was [the best alternative](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/issues/62#issuecomment-707610620).

This [effort of splitting up the CI/CD system](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/blob/main/dev-handbook/ci-cd-split-multiple-workflows.md) took a month. I'm happy to announce that that effort is done now, moving us significantly forward.

## Progress at 79%

At the time of writing, [15 out of 19](https://github.com/fullstaq-labs/fullstaq-ruby-umbrella/blob/main/ROADMAP.md) todo items for epic 4 have been closed. So the progress is at ~79%: epic 4 is close to being finished. Let's press forward and make this project a success, and make sure that it becomes a true community project!

**Comment by patchandpray** [comment]
Awesome work!
